ChainSafe/gossamer
# dot, dot/rpc: implement basic websocket support for RPC

## Changes
Added basic functionality so node can connect via WebSocket for RPC communication.  Uses WebSocket upgrader to forward WebSocket request to RPC server and returns results back to WebSocket connection.



- Creates RPC service that handles WebSocket connections
- 

## Tests:
Currently WebSocket port is set in node config.toml, current default is 8546.  

Need some sort of WebSocket client for testing (let me know if you have tools to suggest), I'm using a browser extension called WebSocket King Client, sort of like Postman but for WebSockets: https://chrome.google.com/webstore/detail/websocket-king-client/cbcbkhdmedgianpaifchdaddpnmgnknn 

Connect to node `ws://localhost:8546/`  and JSON request: For example:
```
{"id":1,"jsonrpc":"2.0","method":"chain_getBlockHash","params":[0]}
```
You can also connect polkadot-js client (although it doesn't do anything yet except acknowledge the connection):
- Browse to https://polkadot.js.org/ 
- Settings: custom endpoint: ws://localhost:8546/
- Start node with --rpc, You'll see the "not connected to node message" disappear.
- Open browser development tools console and you'll see that the client is failing because state_subscribeRuntimeVersion, system_chain and system_properties hasn't been implemented yet.

## Checklist:



- [x] I have read [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md)
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [x] I have linked any relevant issues in my pull request comments
- [x] All integration tests and required coverage checks are passing

## Issues:



- closes #569
